sdk/tools/whitelister/whitelister.cpp

```diff
@@ -0,0 +1,411 @@
+        throw std::runtime_error("popen() failed!");
+    }
+
+    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr)
+    {
+        result += buffer.data();
+    }
+
+    return result;
+}
+
+string
+ExecuteCmd(const char* Format, ...)
+{
+    array<char, _MAX_PATH> buffer;
+    va_list argptr;
+
+    va_start(argptr, Format);
+    _vsnprintf(buffer.data(), buffer.size(), Format, argptr);
+    va_end(argptr);
+
+    return Execute(buffer.data());
+}
+
+string
+git_blame(string &File, int Line)
+{
+    return ExecuteCmd("git blame -L %u,%u --show-number --show-name %s",
+                      Line, Line, File.c_str());
+}
+
+string
+git_rev_blame(string &OldHash, string &OldFile, int OldLine)
+{
+    return ExecuteCmd("git blame -L %u,%u -s --reverse --show-number --show-name %s %s",
+                      OldLine, OldLine, OldHash.c_str(), OldFile.c_str());
+}
+
+string
+git_rev_parse_head(void)
+{
+    string result = ExecuteCmd("git rev-parse @{0}");
+    return result.substr(0, 12);
+}
+
+bool
+git_diff_check(string &File)
+{
+    return ExecuteCmd("git diff --name-only %s", File.c_str()).length() != 0;
+}
+
+int
+AddWhitelistLine(const string &WhitelistFileName, const string &SourceFileName, int Line)
+{
+    return 0;
+}
+
+uint32_t
+GetWhiteListAttributesFromReason(string& Reason)
+{
+    uint32_t attributes = 0;
+
+    if (Reason.find("#ROS") != string::npos)
+    {
+        attributes |= WLA_BROKEN_ROS;
+    }
+    if (Reason.find("#WIN2003") != string::npos)
+    {
+        attributes |= WLA_BROKEN_WIN2003;
+        attributes |= WLA_BROKEN_ROS;
+    }
+    if (Reason.find("#VISTA") != string::npos)
+    {
+        attributes |= WLA_BROKEN_WINVISTA;
+    }
+    if (Reason.find("#WIN7") != string::npos)
+    {
+        attributes |= WLA_BROKEN_WIN7;
+    }
+    if (Reason.find("#WIN8") != string::npos)
+    {
+        attributes |= WLA_BROKEN_WIN8;
+    }
+    if (Reason.find("#WIN81") != string::npos)
+    {
+        attributes |= WLA_BROKEN_WIN81;
+    }
+    if (Reason.find("#WIN10") != string::npos)
+    {
+        attributes |= WLA_BROKEN_WIN10;
+    }
+
+    // When nothing is specified, we assume it's just broken
+    if (attributes == 0)
+    {
+        attributes = WLA_BROKEN_ROS |
+                     WLA_BROKEN_WIN2003 |
+                     WLA_BROKEN_WINVISTA |
+                     WLA_BROKEN_WIN7 |
+                     WLA_BROKEN_WIN8 |
+                     WLA_BROKEN_WIN81 |
+                     WLA_BROKEN_WIN10;
+    }
+
+    Trace("Attributes = 0x%x\n", attributes);
+    return attributes;
+}
+
+void
+AddWhitelistEntry(vector<WL_ENTRY> &Whitelist, string& Line, string& HeadHash, const string& WhiteListFileName)
+{
+    size_t start, length;
+
+    // Format:
+    // <Hash> <original file name> <original line> "<reason>"
+    // 8400fdc0786b rostests/apitests/advapi32/LockDatabase.c 701 "Fails on WHS"
+
+    // Start with the hash, which is 12 characters
+    string oldHash = Line.substr(0, 12);
```

Comment:
This might change if commit history grows, right ?

Reply:
No. The idea is that you use the current commit hash and current line and enter that into the white-list file. Whenever the white-list is processed, this is reverse blamed to get the hash and line in the current code, even if new revisions were added.

Reply:
I mean, the string length (12).

Reply:
No, why would it change? Those are SHA1 hashes, they are 40 hex chars, but this is the default length that git shows. We don't need to be cryptographically secure, so 12 chars should be good enough.

Reply:
Well, I'm gonna switch  to `--porcelain`, which uses the full 40 char hash.